### PR TITLE
convert the QPC ticks to microsecond resolution on Windows Platform

### PR DESCRIPTION
**Pull request by bbi-markslemko**
since play times can exceed 8 hours

The problem: 

> The assert because nTicks is negative: 0xfffffffffe69e34d  
> ```
> 	nTicks = MicroProfileLogTickDifference(nTickStart, nClamped);
> 	MP_ASSERT(nTicks < 0x8000000000000000); 
> ```
> The clamp is calculated from these 2 values which are for MP negative:
> ```
> 	nTickStartLog	0x0000892ecab1f54b 
> 	nTickEndLog	0x0000892ecc6e24d2  
> ```
> on the XBox, we are seeing the counter after loading reach a significant count quickly
> ```nTicks 0x00030043533af511 unsigned __int64```
> after ~15 seconds: 
> ```nTicks 0x00040055baf98a8f unsigned __int64```
> 
> which means the counter will reach capacity at around 8 hours of running or 800  hex * 15 seconds (~30k) 
> 

The solution is to reduce the resolution to sub-microseconds instead of sub-nanoseconds.

Other platforms may need a similar correction. TBD

### CODE REVIEW

File: microprofile.cpp [hunk 1]
```diff
@@ -313,14 +313,22 @@ inline uint64_t MicroProfileGetCurrentThreadId()
...
 	if (nTicksPerSecond == 0)
 	{
 		QueryPerformanceFrequency((LARGE_INTEGER*)&nTicksPerSecond);
+		// reduce the resolution to approximately microseconds - bug: https://dev-mc.visualstudio.com/Badger/_workitems/edit/546030 
+		int muValue = nTicksPerSecond >> 20; // get divisor value for microseconds
+		while (muValue > 1) { 
+			muValue >>= 1;
+			++g_MuBitshiftTicksPerSecond; // determine the log2 of value
+		}
 	}
```
Comment (bbi-davidgalloway) [inline microprofile.cpp:330]: Why not remove this `if` completely? Instead, initialize the nTicksPerSecond in a non 'lazy' style. Just make sure an init is called at the beginning of time.  We call this MicroProfileTicksPerSEcondCpu() often enough that having an unnecessary if

File: microprofile.cpp [hunk 2]
```diff
@@ -330,10 +338,11 @@ inline int64_t MicroProfileGetTick()
 	if (initialTicks == 0)
```
Comment (bbi-davidgalloway) [inline microprofile.cpp:338]: Same here. Why not remove the 'lazy' check for initialTicks == 0 and just do this at the beginning of time only and keep the update uncluttered from an if check that only  happens once in a zillion calls.

File: microprofile.cpp (hunk 1 above)
Comment (BBI-MikeBiddlecombe) [reply microprofile.cpp:330]: I agree, get the frequency once at the beginning and calculate the number of bits of the counter that are not significant to the precision.

**Comment by BBI-MikeBiddlecombe** [comment]
I had to write this out to understand it 

// The max # the ticks can accumulate so that we have have ~usec resolution for 48bits

max_ticks_per_usec = 1_000_000

counters_per_sec = QueryPerformanceFrequency()

counter_to_tick_shift = 0;

while (counters_per_sec > max_ticks_per_usec) {
 counter_to_tick_shift++;
 counters_per_sec >> 1;
}

curr_count = QueryPerformanceCounter();

curr_tick == (init_count - curr_count) >> counter_to_tick_shift;

**Comment by BBI-MikeBiddlecombe** [comment]
You *might* have to update this if the PC goes low powered or something truly cruel. I have never investigated all the rules.

File: microprofile.cpp [hunk 3]
```diff
@@ -313,14 +313,22 @@ inline uint64_t MicroProfileGetCurrentThreadId()
...
 	static int64_t nTicksPerSecond = 0;
 	if (nTicksPerSecond == 0)
 	{
 		QueryPerformanceFrequency((LARGE_INTEGER*)&nTicksPerSecond);
+		// reduce the resolution to approximately microseconds - bug: https://dev-mc.visualstudio.com/Badger/_workitems/edit/546030 
+		int muValue = nTicksPerSecond >> 20; // get divisor value for microseconds
+		while (muValue > 1) { 
+			muValue >>= 1;
+			++g_MuBitshiftTicksPerSecond; // determine the log2 of value
+		}
```
Comment (bbi-davidgalloway) [inline microprofile.cpp:329]: Just for fun you can calculate the log2 of the value as follows without any loop.

int nlz(unsigned k) {
   union {
      unsigned asInt;
      float asFloat;
   };
   int n;

   k = k & ~(k >> 1);           /* Fix problem with rounding. */
   asFloat = (float)k + 0.5f;
   n = 158 - (asInt >> 23);
   return n;
}

muValue = (31 - nlz(nTicksPerSecond>>20));

**Review by bbi-davidgalloway — APPROVED** [review]

File: microprofile.cpp (hunk 3 above)
Comment (bbi-markslemko) [reply microprofile.cpp:329]: ooo clever.

Comment (BBI-MikeBiddlecombe) [reply microprofile.cpp:329]: https://docs.microsoft.com/en-us/cpp/intrinsics/lzcnt16-lzcnt-lzcnt64

**Comment by bbi-markslemko** [comment]
@BBI-MikeBiddlecombe, @bbi-davidgalloway 
I heard you - I actually was thinking about that too. I found the intrinsic I was looking for so, I think this is most efficient now. 
works on all intel 64 bit processors that we need to worry about.

File: microprofile.cpp (hunk 3 above)
Comment (bbi-markslemko) [reply microprofile.cpp]: I believe https://docs.microsoft.com/en-us/cpp/intrinsics/bitscanreverse-bitscanreverse64?view=msvc-170 is more portable

**Comment by bbi-davidgalloway** [comment]
> @BBI-MikeBiddlecombe, @bbi-davidgalloway I heard you - I actually was thinking about that too. I found the intrinsic I was looking for so, I think this is most efficient now. works on all intel 64 bit processors that we need to worry about.

Yeah, that's nice.  :smiley: They are likely using an implementation that takes advantage of the built in SSE4 instruction LZCNT https://en.wikipedia.org/wiki/SSE4#POPCNT_and_LZCNT

**Review by bbi-davidgalloway — APPROVED** [review]

File: microprofile.cpp [hunk 4]
```diff
@@ -1530,6 +1530,10 @@ void MicroProfileInit()
 		mutex.lock();
 	if(bOnce)
 	{
+#if defined(_WIN32)
+		initTicksPerSecond();
+#endif
```
Comment (bbi-davidgalloway) [inline microprofile.cpp:1535]: Is this _WIN32 going to work on XB1 ?

Comment (bbi-markslemko) [reply microprofile.cpp:1535]: yes - tested on scarlett and xb1

**Pull request merged**